### PR TITLE
Simplify the kotlin sample

### PR DESCRIPTION
**Pull request by martinbonnin**
As a epoxy newbie and kotlin fan, I opened the kotlin sample but was a bit confused by a lot of code that, unless I miss something was not strictly needed to get the sample working. I tried to simplify it by making it obvious what comes from "ViewHolder", "Custom View" or "data binding". 

If the code I removed was indeed needed, maybe we could have 2 samples: 1 for simple binding and another one more advanced ?

### CODE REVIEW

**Comment by elihart** [comment]
Hey, thanks for putting in some effort to improve the sample - sorry it was confusing. I do want to keep the existing code though - it is important to have the examples for advanced usages.

Have you seen the wiki? There are several sections on kotlin in the side bar, like https://github.com/airbnb/epoxy/wiki/Kotlin-Model-Examples

I think two separate samples would be confusing, but the existing sample could be updated to have basic examples and advanced usages, clearly marked.

**Comment by martinbonnin** [comment]
Maybe explain a bit more what the advanced usages are ? From the sample in master, I could just remove `ExampleKotlinModels`, `KotlinHolder` and `KotlinModel` and the sample app runs the same (at least seen from the outside). If these classes are useful for performance, architecture or correctness reasons, then I can work on making a "sample selection" screen. As of now they feel to me like the exact boilerplate code I'd like to avoid using epoxy.

Also, I found it a bit disappointing that the EditText has no action ? Or has it ? I was really hopping to see some animation displayed when searching for Lottie components :)

**Comment by elihart** [comment]
The kotlin sample isn't really complete. I started it a while back hoping to actually make it a searchable lottie animation app, but never got around to finishing it. It hasn't been high priority though because kotlin with Epoxy is basically the same as Java, there are only slight tweaks to how the annotations are applied, so having a whole other sample isn't necessary.

If you understand the core concepts of Epoxy, then the kotlin information in the wiki is all you need. The models you deleted aren't "advanced" so much as just normal recommended patterns. The only advanced model would be the one under the "Custom Kotlin Model" section, which is more of a proof of concept suggestion that people could try.

What boilerplate are you talking about? If you want to go the viewholder approach and not create a custom view then the holder pattern is the minimal amount of code needed (unless you use databinding) - the code given in https://github.com/airbnb/epoxy/wiki/Kotlin-Model-Examples is the least amount of code required to do this. The rest of the code lives in the sample module to copy over if you want to use it.

We could create a kotlin module with these patterns and base classes that is officially published as a dependency, but that hasn't been a high priority for me - Kotlin provides a ton of different ways to approach this so I've just given suggestions and let people pick their favorite.

If you want to practice with Epoxy and explore this it would be amazing if you feel like building out the kotlin sample!

**Comment by martinbonnin** [comment]
I see. It's just weird to ship code in a sample app that's not actually used. The sample app relies uniquely on data binding from what I can see so the CustomView/ViewHolder example feel a bit out of place.

I'll try to work on imrpoving this sample these next days.

**Comment by martinbonnin** [comment]
Also, do you have some design documents somewhere ? As a newcomer, I'd love to have an opinionated tutorial. Having multiple options to choose from (ViewHolder, databinding, data classes, custom views) is nice and powerful but is a bit overwhelming when discovering the project. 

Do you mix all these or do you rely on one of them more ? 

As of now, I'm leaning toward using View Holders containing kotlin synthetic properties. They have the advantage of being more "type safe": with them, compilation will fail if some view id is not the the layout XML or is the wrong type, which is not always the case when relying on findViewById

**Comment by elihart** [comment]
The wiki is all of the documentation. It's understandably a lot to take in at first - I haven't created any opinionated tutorial. While I could see that being a bit useful, people generally have their preferences on which system to use and they all work well for different situations so it doesn't matter too much.

Personally we use custom views. This has the benefit of being usable outside of Epoxy if needed, and it is more efficient because on updates only props that have changed are updated, instead of doing a full rebind like with the viewholder.

I see what you're saying about synthetic properties, but honestly having regressions with id mismatches   is not a problem I've seen, even with our dozens of engineers and huge app. Although now we have CI tests that inflate each of our custom views and runs tests on them, so if we did ever have a regression that would catch it.

What you're describing is not a wrong way to go though - like I was saying, it's mostly personal preference

**Comment by martinbonnin** [comment]
I just pushed a new version of the sample apps that shows CustomView, ViewHolders and data classes in action. Let me see what you think

**Comment by elihart** [comment]
Thanks! I'll take a look soon

**Comment by elihart** [comment]
Sorry, I haven't forgotten about this. I did look over it, but haven't had time to go more in depth yet

**Comment by elihart** [comment]
There are several kotlin inspection warnings in `MainActivity` that should be fixed.

Also, since models are only built on tab selected nothing is drawn the first time.

Lastly, the tab design looks crunched up with not enough space. Can you change it to just show all of the models in a single vertical list?

image

image

File: kotlinsample/src/main/java/com/airbnb/epoxy/kotlinsample/MainActivity.kt [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+                    2 -> showCustomViewItems()
+                    3 -> showViewHolder()
+                    4 -> showDataClass()
+                }
+            }
+        })
+    }
+
+    private fun showMixAndMatch() {
+        recyclerView.buildModelsWith {
```
[the remark this answers is not in the document]
Comment (martinbonnin) [reply kotlinsample/src/main/java/com/airbnb/epoxy/kotlinsample/MainActivity.kt]: Yep, I did not do it so as not to add more code to the project. Any reason this is not included in the core lib ? Maybe just generate it along the other extensions ?

File: kotlinsample/src/main/java/com/airbnb/epoxy/kotlinsample/models/ItemCustomView.kt [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    attrs: AttributeSet? = null,
+    defStyleAttr: Int = 0
+) : LinearLayout(context, attrs, defStyleAttr) {
+
+    init {
+        inflate(context, R.layout.custom_view_item, this)
+        orientation = VERTICAL
+    }
+
+    lateinit var greeting: String
```
[the remark this answers is not in the document]
Comment (martinbonnin) [reply kotlinsample/src/main/java/com/airbnb/epoxy/kotlinsample/models/ItemCustomView.kt]: I know, I still wanted to keep this method as an example. I'll add more comments

**Comment by martinbonnin** [comment]
Sample updated. Let me know what you think.

File: kotlinsample/src/main/java/com/airbnb/epoxy/kotlinsample/MainActivity.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (elihart) [reply kotlinsample/src/main/java/com/airbnb/epoxy/kotlinsample/MainActivity.kt]: I've avoided adding the kotlin stdlib dependency to the core library, but it's something I'll probably do eventually. I could also ship a kotlin support module, just hasn't been high priority. Generating them is an interesting idea, but having the code not exist until build can be annoying

**Comment by elihart** [comment]
@martinbonnin this is great! thanks for the cleanup and for working on this

File: kotlinsample/src/main/java/com/airbnb/epoxy/kotlinsample/MainActivity.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (martinbonnin) [reply kotlinsample/src/main/java/com/airbnb/epoxy/kotlinsample/MainActivity.kt]: > having the code not exist until build can be annoying

Yup, this is my main gripe against annotation processors. In that case though, intellij will already show red everywhere for the model extensions so having one more line red might not be that problematic.

**Pull request merged**